### PR TITLE
Feature/nodes metadata

### PR DESCRIPTION
**Pull request by Sypheos**
This is in response of #507 

Implementation of built-in attributes for devices.

It is now possible to set builtin for a device as the form of a key-value. They are store as a slice alphabetically ordered by keys.
The feature is added to the grpc API, new options are added to ttnctl 'attr-add' and 'attr-remove'

### CODE REVIEW

File: api/handler/handler.proto [hunk 1]
```diff
@@ -5,6 +5,7 @@ syntax = "proto3";
 
 import "google/protobuf/empty.proto";
 import "google/api/annotations.proto";
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
```
Comment (htdvisser) [inline api/handler/handler.proto]: Not needed

File: cmd/handler.go [hunk 2]
```diff
@@ -112,6 +112,13 @@ var handlerCmd = &cobra.Command{
 		} else {
 			ctx.Warn("AMQP is not enabled in your configuration")
 		}
+
+		if viper.GetString("handler.device-builtin") != "" {
```
Comment (htdvisser) [inline cmd/handler.go]: Shouldn't the ttn builtin attributes be hardcoded in the binary? We can specify additional attributes with the `--extra-device-attributes` config

File: cmd/handler.go [hunk 3]
```diff
@@ -112,6 +112,13 @@ var handlerCmd = &cobra.Command{
...
+			handler = handler.SetDeviceBuiltinList(viper.GetString("handler.device-builtin"))
+		} else {
+			ctx.Warn("No builtin attribute set in your configuration")
```
Comment (htdvisser) [inline cmd/handler.go]: Since when is that a problem?

File: cmd/handler.go [hunk 4]
```diff
@@ -210,4 +217,7 @@ func init() {
 	handlerCmd.Flags().Int("http-port", 8084, "The port where the gRPC proxy should listen")
 	viper.BindPFlag("handler.http-address", handlerCmd.Flags().Lookup("http-address"))
 	viper.BindPFlag("handler.http-port", handlerCmd.Flags().Lookup("http-port"))
+
+	handlerCmd.Flags().String("builtin-attr", "", "BuiltIn attributes list for a device")
```
Comment (htdvisser) [inline cmd/handler.go]: Inconsistent with `device-builtin` that you use above

File: core/handler/device/converter.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+package device
+
+import (
+	pb "github.com/TheThingsNetwork/ttn/api/handler"
+	pb_lorawan "github.com/TheThingsNetwork/ttn/api/protocol/lorawan"
+)
+
+//Convert a device representation to a protoBuffer device struct
```
Comment (htdvisser) [inline core/handler/device/converter.go]: Missing space after `//`

File: core/handler/device/converter.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			ActivationConstraints: dev.Options.ActivationConstraints,
+		}},
+		Latitude:  dev.Latitude,
+		Longitude: dev.Longitude,
+		Altitude:  dev.Altitude,
+		Builtin:   dev.Builtin,
+	}
+}
+
+//FromPb create a new Device from a protoBuffer Device
```
Comment (htdvisser) [inline core/handler/device/converter.go]: - Missing space after `//`. Please check for this in other places as well
- create**s**. Also check this in other places

File: core/handler/device/converter.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+		dev.DevAddr = *lorawan.DevAddr
+	}
+	if lorawan.NwkSKey != nil {
+		dev.NwkSKey = *lorawan.NwkSKey
+	}
+	if lorawan.AppSKey != nil {
+		dev.AppSKey = *lorawan.AppSKey
+	}
+	if lorawan.AppKey != nil {
+		if dev.AppKey != *lorawan.AppKey { // When the AppKey of an existing device is changed
```
Comment (htdvisser) [inline core/handler/device/converter.go]: This doesn't belong in the conversion, but in the manager

File: core/handler/device/converter_test.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
+// Copyright © 2017 The Things Network
+// Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+
+package device
+
+import (
+	"testing"
+
```
Comment (htdvisser) [inline core/handler/device/converter_test.go]: Format imports

File: core/handler/device/converter_test.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+var test_dev = &Device{
+	AppEUI: [8]byte{0x10},
+	AppID:  "test-app",
+	DevEUI: [8]byte{0x10},
+	DevID:  "test-dev",
+
+	Description: "testing",
+	Latitude:    255,
```
Comment (htdvisser) [inline core/handler/device/converter_test.go]: Please use valid (or at least plausible) values

File: core/handler/device/converter_test.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	Longitude:   255,
+	Altitude:    255,
+
+	Options: Options{ActivationConstraints: "activate"},
```
Comment (htdvisser) [inline core/handler/device/converter_test.go]: Check the documentation for `ActivationConstraints`

File: core/handler/device/converter_test.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+	AppKey:        [16]byte{0x10},
+	UsedDevNonces: []DevNonce{},
+	UsedAppNonces: []AppNonce{},
+
+	DevAddr: types.DevAddr{byte('E')},
```
Comment (htdvisser) [inline core/handler/device/converter_test.go]: Inconsistent. Should probably be `[4]byte{0x10}`

File: core/handler/device/device.go [hunk 12]
```diff
@@ -7,6 +7,9 @@ import (
 	"reflect"
 	"time"
 
+	"sort"
```
Comment (htdvisser) [inline core/handler/device/device.go]: Format imports

File: core/handler/device/device.go [hunk 13]
```diff
@@ -7,6 +7,9 @@ import (
...
+
+	pb "github.com/TheThingsNetwork/ttn/api/handler"
```
Comment (htdvisser) [inline core/handler/device/device.go]: We usually call it `pb_handler`

File: core/handler/device/device.go [hunk 14]
```diff
@@ -54,6 +57,8 @@ type Device struct {
 
 	CreatedAt time.Time `redis:"created_at"`
 	UpdatedAt time.Time `redis:"updated_at"`
+
+	Builtin []*pb.Attribute `redis:"builtin"`
```
Comment (htdvisser) [inline core/handler/device/device.go]: I really don't like the term `Builtin`, why not call it `Attributes`?

File: core/handler/device/device_test.go [hunk 15]
```diff
@@ -52,3 +53,92 @@ func TestDeviceGetLoRaWAN(t *testing.T) {
 	}
 	device.GetLoRaWAN()
 }
+
+func TestDevice_MapOldBuiltin(t *testing.T) {
+	a := New(t)
+
+	device := &Device{
+		old: nil,
```
Comment (htdvisser) [inline core/handler/device/device_test.go]: old is nil by default

File: core/handler/device/store.go [hunk 16]
```diff
@@ -122,6 +130,7 @@ func (s *RedisDeviceStore) Set(new *Device, properties ...string) (err error) {
 	if new.old == nil {
 		new.CreatedAt = now
 	}
+	s.builtinFilter(new)
```
Comment (htdvisser) [inline core/handler/device/store.go]: Yeah definitely rename it to attributes, `builtinFilter` does not make _any_ sense

File: cmd/handler.go (hunk 3 above)
Comment (Sypheos) [inline cmd/handler.go]: Change to `Info` ? At least notify that no builtin are set in the config

File: cmd/handler.go (hunk 2 above)
Comment (Sypheos) [inline cmd/handler.go]: There is this option. Add permanent attributes in the binary can ensure some attribute keys will always be accepted.
Changing the config to `--extra-devices-attributes` or `--devices-extra-attributes` ?

File: core/handler/device/store.go [hunk 17]
```diff
@@ -122,6 +139,24 @@ func (s *RedisDeviceStore) Set(new *Device, properties ...string) (err error) {
 	if new.old == nil {
 		new.CreatedAt = now
 	}
+	customAttributeSlots := maxDeviceAttributes
+	for k, v := range new.Attributes {
+		if idx := sort.SearchStrings(s.allowedAttributes, k); idx < len(s.allowedAttributes) && s.allowedAttributes[idx] == k {
+			fmt.Println("Setting attribute", k, v)
```
Comment (johanstokking) [inline core/handler/device/store.go]: Remove or change to log

File: core/handler/device/store.go [hunk 18]
```diff
@@ -122,6 +139,24 @@ func (s *RedisDeviceStore) Set(new *Device, properties ...string) (err error) {
...
+		if len(k) > maxDeviceAttributeKeyLength {
+			return fmt.Errorf(`Attribute key "%s" exceeds maximum length (%d)`, k, maxDeviceAttributeKeyLength)
+		}
+		if len(v) > maxDeviceAttributeValueLength {
+			return fmt.Errorf(`Attribute value for key "%s" exceeds maximum length (%d)`, k, maxDeviceAttributeValueLength)
+		}
+		if customAttributeSlots < 1 {
+			return fmt.Errorf(`Maximum number of custom attributes (%d) exceeded`, maxDeviceAttributes)
+		}
+		fmt.Println("Setting attribute", k, v)
```
Comment (johanstokking) [inline core/handler/device/store.go]: Idem

File: core/handler/device/store.go (hunk 17 above)
Comment (htdvisser) [inline core/handler/device/store.go]: shouldn't have been committed. sry

File: core/handler/device/store.go (hunk 18 above)
Comment (htdvisser) [inline core/handler/device/store.go]: idem

**Pull request merged**